### PR TITLE
Only evaluate julia lazily in flake.nix

### LINKED ISSUE #410: derivation for julia is evaluated on each kernel
When trying to enable support aarch64-darwin I stumbled upon this. On each build of any example kernel, the Julia package will be evaluated while there is no need for that actually. In my case an error will be thrown it;s not supported on my system. See linked PR for proposed fix 
```
nix build .\#jupyterlab-kernel-example-zsh-minimal
warning: Git tree '/Users/jaapterwoerds/workspaces/jaapterwoerds/tmp/jupyterWith' is dirty
warning: Using saved setting for 'extra-substituters = https://tweag-jupyter.cachix.org' from ~/.local/share/nix/trusted-settings.json.
warning: Using saved setting for 'extra-trusted-public-keys = tweag-jupyter.cachix.org-1:UtNH4Zs6hVUFpFBTLaA4ejYavPo5EFFqgd7G7FxGW9g=' from ~/.local/share/nix/trusted-settings.json.
error: Package ‘julia-bin-1.8.3’ in /nix/store/n0f21zpy9w90sa9v75iwx6fmk897z3nw-source/pkgs/development/compilers/julia/1.8-bin.nix:56 is not supported on ‘aarch64-darwin’, refusing to evaluate.

       a) To temporarily allow packages that are unsupported for this system, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1

        Note: For `nix shell`, `nix build`, `nix develop` or any other Nix 2.4+
        (Flake) command, `--impure` must be passed in order to read this
        environment variable.

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowUnsupportedSystem = true; }
       in configuration.nix to override this.

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowUnsupportedSystem = true; }
       to ~/.config/nixpkgs/config.nix.
(use '--show-trace' to show detailed location information)
```

https://github.com/tweag/jupyterWith/blob/2bbd18ded31c5a27c86e6bd2aae2c2d79e7d180c/flake.nix#L436

### PR DESCRIPTION
**Pull request by jaapterwoerds**
fixes #410

### CODE REVIEW

**Comment by djacu** [comment]
Hey can you run `alejandra` on the repo and push a new commit? The CI is failing because of the linter.
`alejandra` is in the devShell so you can run

```
nix develop
pre-commit run --all
```

File: flake.nix [hunk 1]
```diff
@@ -416,34 +416,35 @@
...
+                    --set JUPYTER_PATH ${lib.concatStringsSep ":" kernelDerivations} \
+                    --set JUPYTER_CONFIG_DIR "${jupyterDir}/config" \
+                    --set JUPYTER_DATA_DIR ".jupyter/data" \
+                    --set IPYTHONDIR "/path-not-set" \
+                    --set JUPYTER_RUNTIME_DIR ".jupyter/runtime"
                 done
-              fi
-            '';
+              ''
+              + (lib.strings.optionalString ((lib.lists.count (e: builtins.match ".*julia.*" e.outPath != null) kernelDerivations) > 0) ''
```
Comment (djacu) [inline flake.nix:436]:
```suggestion
              + (lib.strings.optionalString (
                  builtins.any
                  (kernel: kernel.kernelInstance.language == "julia")
                  kernelDerivations
                )
 ''
```

This should be a bit safe than looking in the outpath of the kernel derivation.

File: flake.nix [hunk 2]
```diff
@@ -416,34 +416,35 @@
...
+                # add Julia for IJulia
```
Comment (djacu) [inline flake.nix:437]: Can you modify the scripts that add Julia to be the following. We don't need the `if` statement any more.
```
# add Julia for IJulia
echo 'Adding Julia as an available package.'
for i in ${pkgs.julia-bin}/bin/*; do
  filename=$(basename $i)
  ln -s ${pkgs.julia-bin}/bin/$filename $out/bin/$filename
done
```

Comment (djacu) [inline flake.nix:437]: Can you modify the scripts that add Julia to be the following. We don't need the `if` statement any more.
```
# add Julia for IJulia
echo 'Adding Julia as an available package.'
for i in ${pkgs.julia-bin}/bin/*; do
  filename=$(basename $i)
  ln -s ${pkgs.julia-bin}/bin/$filename $out/bin/$filename
done
```

**Comment by jaapterwoerds** [comment]
Great suggestions I will update the MR, thanks

**Comment by jaapterwoerds** [comment]
Sorry for the force push I wasn't aware of the other commits made. I hope it's Ok now

**Comment by djacu** [comment]
> Sorry for the force push I wasn't aware of the other commits made. I hope it's OK now

No worries. I had just caught the branch up to main so the CI would run properly. Looks like you did the same.

File: flake.nix [hunk 3]
```diff
@@ -416,34 +416,35 @@
...
+                  (kernel: kernel.kernelInstance.language == "julia")
+                  kernelDerivations
+                ) ''
+                  # add Julia for IJulia
+                  echo 'Adding Julia as an available package.'
+                  for i in ${pkgs.julia-bin}/bin/*; do
+                    filename=$(basename $i)
+                    ln -s ${pkgs.julia-bin}/bin/$filename $out/bin/$filename
+                  done
+                '')));
```
Comment (djacu) [inline flake.nix:447]:
```suggestion
              + (lib.strings.optionalString (
                  builtins.any
                  (kernel: kernel.kernelInstance.language == "julia")
                  kernelDerivations
                ) ''
                  # add Julia for IJulia
                  echo 'Adding Julia as an available package.'
                  for i in ${pkgs.julia-bin}/bin/*; do
                    filename=$(basename $i)
                    ln -s ${pkgs.julia-bin}/bin/$filename $out/bin/$filename
                  done
                ''));
```

Looks like `optionalString` got accidentally doubled up here.

**Review by djacu — CHANGES_REQUESTED** [review]

File: flake.nix (hunk 3 above)
Comment (djacu) [reply flake.nix:447]: Hey I'm going to sign off on this so we can get the PR merged and unblock you.

**Review by djacu — APPROVED** [review]
lgtm

**Comment by jaapterwoerds** [comment]
Thanks for the fix @djacu

**Pull request merged**